File: src/main/GO/leetcode/lists/topinterview/LeetCode329.go
package topinterview

//https://leetcode.com/problems/longest-increasing-path-in-a-matrix/solutions/2054922/java-3-approaches-dfs-bfs-memoization-dfs/
//https://leetcode.cn/problems/longest-increasing-path-in-a-matrix/solution/tong-ge-lai-shua-ti-la-yi-ti-si-jie-bfs-agawl/

//https://leetcode.com/problems/longest-increasing-path-in-a-matrix/solutions/775043/both-dfs-and-bfs-solutions/

var dir = [][]int{{1, 0}, {0, 1}, {-1, 0}, {0, -1}}

func longestIncreasingPath(matrix [][]int) int {
	if len(matrix) == 0 {
		return 0
	}
	res := 0
	indegree := make([][]int, len(matrix))
	for i := 0; i < len(indegree); i++ {
		indegree[i] = make([]int, len(matrix[0]))
	}
	q := make([][]int, 0)

	for i := 0; i < len(matrix); i++ {
		for j := 0; j < len(matrix[0]); j++ {
			indegreeCount := 0
			for k := 0; k < 4; k++ {
				r := i + dir[k][0]
				c := j + dir[k][1]
				if r >= 0 && c >= 0 && r < len(matrix) && c < len(matrix[0]) && matrix[r][c] < matrix[i][j] {
					indegreeCount++
				}
			}
			indegree[i][j] = indegreeCount
			if indegree[i][j] == 0 {
				q = append(q, []int{i, j, 1})

			}
		}
	}
	for len(q) > 0 {
		node := q[0]
		q = q[1:]
		x := node[0]
		y := node[1]
		step := node[2]
		res = Max(res, step)
		for k := 0; k < 4; k++ {
			r := x + dir[k][0]
			c := y + dir[k][1]
			if r >= 0 && c >= 0 && r < len(matrix) && c < len(matrix[0]) && matrix[r][c] > matrix[x][y] {
				indegree[r][c]--
				if indegree[r][c] == 0 {
					q = append(q, []int{r, c, step + 1})

				}
			}
		}
	}
	return res
}

File: java/src/leetcode/problems/LeetCode329.java
package leetcode.problems;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/11/16 18:08
 */
public class LeetCode329 {


    int[][] memo;

    public int longestIncreasingPath(int[][] matrix) {
        if (matrix == null) {
            return 0;
        }
        int m = matrix.length;
        int n = matrix[0].length;
        memo = new int[m][n];
        int res = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                res = Math.max(res, dfs(matrix, i, j));
            }
        }
        return res;
    }

    int[][] dirs = new int[][]{{-1, 0}, {1, 0}, {0, 1}, {0, -1}};

    private int dfs(int[][] matrix, int row, int col) {
        if (memo[row][col] != 0) {
            return memo[row][col];
        }
        int res = 1;
        for (int i = 0; i < dirs.length; i++) {
            int xn = row + dirs[i][0];
            int yn = col + dirs[i][1];
            if (xn < 0 || xn >= matrix.length || yn < 0 || yn >= matrix[0].length || matrix[xn][yn] <= matrix[row][col]) {
                continue;
            }
            res = Math.max(res, dfs(matrix, xn, yn) + 1);
        }
        memo[row][col] = res;
        return res;
    }
}
